src/core/thread/src_match_controller.hpp

```diff
@@ -0,0 +1,204 @@
+ *
+ * @brief
+ *   This module includes definition for source address match controller.
+ *
+ * @{
+ */
+
+/**
+ * This class implements the "source address match" controller.
+ *
+ * The source address match feature controls how the radio layer decides the "frame pending" bit for acks sent in
+ * response to data request commands from sleepy children.
+ *
+ * This class updates the source match table and also controls when to enable or disable the source matching
+ * feature.
+ *
+ * The source address match table provides the list of children for which there is a pending frame. Either a short
+ * address or an extended/long address can be added to the source address match table.
+ *
+ * If enabled, the radio layer uses the source address match table to determine whether to set or clear the "frame
+ * pending" bit in an ack to a data request command. If disabled, the radio layer must set the "frame pending" on all
+ * acks to data request commands.
+ *
+ */
+class SourceMatchController
+{
+public:
+    /**
+     * This constructor initializes the object.
+     *
+     * @param[in]  aMeshForwarder   A reference to the Mesh Forwarder.
+     *
+     */
+    explicit SourceMatchController(MeshForwarder &aMeshForwarder);
+
+    /**
+     * This method returns the pointer to the parent otInstance structure.
+     *
+     * @returns The pointer to the parent otInstance structure.
+     *
+     */
+    otInstance *GetInstance(void);
+
+    /**
+     * This method returns the current state of source address matching.
+     *
+     * @returns `true` if source address matching is enabled, `false` otherwise.
+     *
+     */
+    bool IsEnabled(void) const { return mEnabled; }
+
+    /**
+     * This method increases the message count for a child and updates the source match table.
+     *
+     * @param[in] aChild    A reference to the child.
+     *
+     */
+    void IncreaseMessageCount(Child &aChild);
+
+    /**
+     * This method decreases the message count for a child and updates the source match table.
+     *
+     * @param[in] aChild    A reference to the child.
+     *
+     */
+    void DecreaseMessageCount(Child &aChild);
+
+    /**
+     * This method resets the message count for a child to zero and updates the source match table.
+     *
+     * @param[in] aChild    A reference to the child.
+     *
+     */
+    void ResetMessageCount(Child &aChild);
+
+    /**
+     * This method sets whether or not to perform source address matching on the extended or short address for
+     * a child.
+     *
+     * @param[in] aChild            A reference to the child.
+     * @param[in] aUseShortAddress  `true` to match on short source address, `false` otherwise.
+     *
+     */
+    void SetSrcMatchAsShort(Child &aChild, bool aUseShortAddress);
+
+private:
+    /**
+     * This method clears the source match table.
+     *
+     */
+    void ClearTable(void);
+
+    /**
+     * This method enables or disables the source matching..
```

Comment:
nit: double `..`

Also, indicate that when source matching is disabled, the Frame Pending bit is set in all IEEE 802.15.4 ACKs to IEEE 802.15.4 Data Request messages.